[CZBANK-14] feature: create, delete, rename bank account via UI

silviczka opened the pull request:
Created CreateDialog UI component for New bank account creation using shadcn/ui components. 
This PR covers the UI only; backend integration will follow separately (CZBANK-14)

silviczka committed:
subtask: create dialog component for new bank account UI

silviczka commented:
Hello, I'm not sure if I understood the ticket instructions correctly, please check my code when you have time and let me know what to change. Thx!

vojtech-cerveny commented:
Yeah, this is exactly what I wanted. 

It would be nice to create new "row" for buttons and don't have it in same row as bank account cards. 

Also - please for currency - right now, we support only CZECHITOKEN, so please replace text input for Select - with only one value - CZECHITOKEN.

<img width="1031" height="580" alt="image" src="https://github.com/user-attachments/assets/8909c713-854d-4359-a62f-932341e7b212" />


And if you want - you can create form validation - we have some form, so it should be "copy&paste". Or add backend part. It is up to you, how do you want to continue. :)

Otherwise 
<img width="300" height="300" alt="image" src="https://github.com/user-attachments/assets/cd2d4225-249b-4bbf-852c-056f6ebf900f" />

vojtech-cerveny commented on src/app/page.tsx:
```diff
@@ -23,6 +24,7 @@ export default async function Home() {
             <BankAccountCard bankAccount={ba} />
           </Link>
         ))}
+        <CreateDialog />
```
Something like 
```
      <div className="flex justify-end mb-4">
        <CreateDialog />
      </div>
```

silviczka replied:
done, now it is on right side, is that right?

vojtech-cerveny commented on src/components/bank-account/create-dialog.tsx:
```diff
@@ -0,0 +1,32 @@
+  return (
+    <Dialog>
+      <DialogTrigger asChild>
+        <Button>Create new</Button>
```
Can you add some icon here as well? We have some icons from package `lucide-react` - https://lucide.dev/icons/ , so just pick one and import it there :)

silviczka replied:
done

vojtech-cerveny commented on src/components/bank-account/create-dialog.tsx:
```diff
@@ -0,0 +1,32 @@
+          </div>
+          <div className="space-y-2">
+            <Label htmlFor="currency">Currency</Label>
+            <Input id="currency" placeholder="e.g. USD" />
```
Use Select here please

silviczka replied:
done

vojtech-cerveny reviewed (commented):
Consider my suggestions please :)

silviczka committed:
chore: update readme setup

silviczka committed:
chore: change btn position, replace input with select

silviczka commented:
I think I've covered all the more specific suggestions in my last commit

Next I will focus on >creation of form validation, adding backend part, but I think it might be the CZBANK-14, should I work on this in CZBANK-54?

Pls check if position of the btn is correct now :) 
THX

postulkam commented on src/components/bank-account/create-dialog.tsx:
```diff
@@ -0,0 +1,44 @@
+      </DialogContent>
+    </Dialog>
+  );
+}
```
Should we consider making these two from inputs as required?

postulkam commented on src/components/bank-account/create-dialog.tsx:
```diff
@@ -0,0 +1,44 @@
+                <SelectValue placeholder="Select a currency" />
+              </SelectTrigger>
+              <SelectContent>
+                <SelectItem value="CZECHITOKEN">CZECHITOKEN</SelectItem>
```
There is only one currency for now, can we somehow always automatically pre-select the CZECHITOKEN value?

vojtech-cerveny replied:
`Select` should have `defaultValue` or something like that. Btw. shadcn uses `radix primitives` - you can check their documentation, I didn't find anything useful on shadcn docs. 
https://www.radix-ui.com/primitives/docs/overview/introduction

silviczka replied:
defaultValue implemented in last commit, I'm gonna focus on BE in next commits

postulkam reviewed (dismissed):
Do we really want the button on the right side? 

<img width="1510" height="383" alt="image" src="https://github.com/user-attachments/assets/2930627e-adb2-45ae-ac55-16779466d478" />
<img width="1510" height="383" alt="image" src="https://github.com/user-attachments/assets/912a16c5-03e8-4059-9f62-e087bd0e2ff8" />

----

The code is perfect, I have two questions and one UI proposal to consider to make. You can resolve them and decide not to work on them, it is perfectly ok with me. :)

postulkam commented:
> 
> Next I will focus on >creation of form validation, adding backend part, but I think it might be the CZBANK-14, should I work on this in CZBANK-54?
> 

Either way is ok in this situation, this is not some breaking change.
But continue to work on this branch is better IMHO. The dialog will not run correctly without the BE part and we usually do not merge something that leads user to death end to develop branch. 

Normally, in like 90 percent of the time, the BE part will be done first by BE developers - that can be merged without problems, user cannot "touch" or interact with it. You will be given some info about the endpoints and then you will create the dialog and connect the BE and FE part.

silviczka commented:
just to reply to this>Do we really want the button on the right side? 

I was also a bit confused, cause I thought we said on call it would be above the card, but then I checked the comments and @vojtech-cerveny wrote he wanted flex justify-end

Not sure why the original comment isn't here, though, ...I'm open to any changes :)

vojtech-cerveny reviewed (dismissed):
Action button on the right side is OK for me and better than left side. 

LGTM, please name commit message as `feature: [CZBANK-54] create dialog component for new bank account UI` , but you can expand this PR with backend part :)

silviczka committed:
Merge remote-tracking branch 'origin/develop' into CZBANK-54

silviczka committed:
fix: set CZECHITOKEN as default currency

silviczka committed:
feature: connect frontend form to backend API with cookie auth

silviczka commented:
Hello, 
I have tried to implement the backend part. I was not able to access apikey in backend in order to use it for validation. That's why I used cookie validation only.

I made a small change to env.ts as the discord hook was breaking the transfer page.

Please let me know what to change, add, remove. Or where to access the apikey in backend :D
Thanks

silviczka committed:
fix: apikey validation added

postulkam commented on src/app/page.tsx:
```diff
@@ -9,19 +9,32 @@ import Link from "next/link";
+            ipAddress,
+            userAgent,
+          }}
+        />
```
I would pass a whole session here, I think, it is not necessary to deconstruct it.

postulkam commented on src/components/bank-account/create-dialog.tsx:
```diff
@@ -16,19 +16,31 @@ import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@
+interface CreateDialogProps {
+  session: Session;
+}
+
```
Nice!

postulkam commented on src/components/bank-account/create-dialog.tsx:
```diff
@@ -53,7 +65,7 @@ export function CreateDialog() {
         variant: "destructive",
         title: "Failed to create account",
-        description: response.error?.message || "Unknown error",
+        description: getErrorMessage(response.error),
```
Good idea of moving this into specific function, it is more clear now.

postulkam commented on src/domain/bankAccount-domain/ba-actions.ts:
```diff
@@ -1,20 +1,28 @@
+  return validCurrencies.includes(value as Currency);
+}
+
+export async function createBankAccountAction(data: { name: string; currency: string }, session: Session) {
```
I noticed some TypeScript in this PR. :)

You can add types to functions return value, here, it will be 

```
export async function createBankAccountAction(data: { name: string; currency: string }, session: Session): Promise<void> {
```

postulkam commented on src/domain/bankAccount-domain/ba-actions.ts:
```diff
@@ -1,20 +1,28 @@
+export async function createBankAccountAction(data: { name: string; currency: string }, session: Session) {
+  try {
+    if (!isValidCurrency(data.currency)) {
+      return { success: false, error: `Invalid currency: ${data.currency}` };
```
Look at it, I am not sure that we can use it or not.

In the app, we have something like `successResponse` and `errorResponse`, can we use it here?

postulkam reviewed (commented):
Few questions, please look at it. :)

silviczka committed:
feature: delete ba btn added, rerender on state change, chores based on CR

silviczka commented:
Hello, 
thanks for CR, I have added DeleteBtn and implemented rerender of BA list + done chores from CR suggestions. (at least I hope I did it right)

Can you please check the newest commit when anyone of you have time?

vojtech-cerveny commented on src/app/api/v1/bank-account/delete/route.ts:
```diff
@@ -0,0 +1,30 @@
+import { checkUserAuthOrThrowError } from "@/app/api/v1/server-actions";
```
This comment is for whole file - in general this is great and we need just couple changes here :)

I would love to move this into `api/v1/bank-account/{id}/route.ts` - and as you can see, we already have this there :)) I noticed it right now (I forgot :D that we have it).

So compare your solution and existing solution. And we should have URL for deleting under {id}, when URL should be `DELETE api/v1/bank-account/fe123xxa` rather than `bank-account/delete?fe123xxa`

vojtech-cerveny commented on src/app/api/v1/bank-account/delete/route.ts:
```diff
@@ -0,0 +1,30 @@
+    }
+    const result = await bankAccountService.deleteBankAccount(bankAccountId);
+
+    await bankAccountService.deleteBankAccount(bankAccountId);
```
This is unnecessary - this will throw error everytime, because you are doing this operation on line 18 and here on line 20 :) 
```
line 18 - delete it once
line 20 - trying to delete it again, but it is not in DB -> error
```

vojtech-cerveny commented on src/app/page.tsx:
```diff
@@ -1,49 +1,31 @@
 
 export default async function Home() {
-  const session = await userService.server.getSession(await headers());
+  const apikeyHeaders = await headers();
```
I don't think that name `apiKeyHeaders` - it is storing all headers and it is not connected to apiKey at all (I think).

vojtech-cerveny commented on src/components/bank-account/ba-card.tsx:
```diff
@@ -0,0 +1,41 @@
+  onDelete?: () => void;
+}
+
+export default function BankAccountCard({ bankAccount, session, onDelete }: Props) {
```
Now `Card` is not wrapped by `Link` - it is not clicable - and now you can't send transactions, because you can't access the bank account detail :)) Can you please wrap this into `Link` component?

vojtech-cerveny commented on src/app/page.tsx:
```diff
@@ -1,49 +1,31 @@
   if (!session) {
     redirect("/signin");
   }
+  const { id, token, userId, expiresAt, createdAt, updatedAt, ipAddress, userAgent } = session.session;
```
All destructured elements are unused. ts(6198)

vojtech-cerveny commented on src/domain/bankAccount-domain/ba-actions.ts:
```diff
@@ -0,0 +1,58 @@
+  }
+}
+
+export async function deleteBankAccountAction(
```
When I tried to remove bank account with transactions and I got this error - 

```
Invalid `__TURBOPACK__imported__module__$5b$project$5d2f$src$2f$lib$2f$db$2e$ts__$5b$app$2d$rsc$5d$__$28$ecmascript$29$__["default"].bankAccount.delete()` invocation in
/Users/vojtechcc/Documents/dev/czechitas/czechibank/.next/server/chunks/ssr/[root-of-the-server]__f7d1e6bd._.js:263:173

  260     return bankAccount;
  261 }
  262 async function deleteBankAccount(bankAccountId) {
→ 263     const bankAccount = await __TURBOPACK__imported__module__$5b$project$5d2f$src$2f$lib$2f$db$2e$ts__$5b$app$2d$rsc$5d$__$28$ecmascript$29$__["default"].bankAccount.delete(
Foreign key constraint violated: `Transaction_fromBankId_fkey (index)`
```

**The Problem**: When you try to delete a bank account, Prisma prevents the deletion because there are transactions that reference that account as either the sender (fromBankId) or receiver (toBankId). This is a database integrity protection - you can't delete a bank account that has transaction history.

```prisma
model Transaction {
  fromBankId String
  toBankId   String
  from       BankAccount @relation("FromAccount", fields: [fromBankId], references: [id])
  to         BankAccount @relation("ToAccount", fields: [toBankId], references: [id])
}
```


Hmmm, I found out this right now, not before. This is bug which is loong time in the code. 

I suggest to create new column in DB - in prisma `isDeleted` - default value (false) and when we will delete bankAccount, we will remove it from list and then we will return everywhere bankAccounts which has `isDeleted == false` 
Or `isActive` if this is better naming.

silviczka replied:
Thanks for review! I will work on all changes including the bug 🪲

vojtech-cerveny reviewed (commented):
In general this is great and you did great job. I also have some comments and I found some issues. One issue is not related to your work directly, so it is up to you if you want to fix it in this PR or we can log as BUG and fix it later. - bug with deleting account with transactions.

Let me know :)

postulkam commented on src/app/api/v1/bank-account/create/route.ts:
```diff
@@ -50,13 +51,13 @@ export async function POST(request: Request) {
 
     if (!parsedBody.success) {
-      return Response.json(errorResponse(parsedBody.error.message, ApiErrorCode.VALIDATION_ERROR));
+      return NextResponse.json(errorResponse(parsedBody.error.message, ApiErrorCode.VALIDATION_ERROR), { status: 400 });
```
Did you added the `400` code to swagger documentation in this file?

postulkam commented on src/app/page.tsx:
```diff
@@ -1,49 +1,31 @@
+          userId: session.session.userId,
+          name: session.user.name,
+        }}
+      />
```
As @vojtech-cerveny mentioned. 

You are destructing the session object, but you don't need that and here, you are using the `session.session` again.

You have two options:
1, use `session.session` here and do not crate this `const { id, token, userId, expiresAt, createdAt, updatedAt, ipAddress, userAgent } = session.session;`

2, use only needed parameters 
```
const { id, token } = session.session;
```
and here use 
```
token: token,
userId: id,
```

For this situation, the 1 is clearer, because you don't create unnecessary constant and you are not using it in more places in this file.

postulkam commented on src/components/bank-account/ba-list.tsx:
```diff
@@ -0,0 +1,40 @@
+interface Props {
+  initialBankAccounts: BankAccount[];
+  session: Session;
+}
```
If you want
- here, you created the Props interface
- you cen create them in other files too :)

postulkam commented on src/components/bank-account/delete-ba-button.tsx:
```diff
@@ -0,0 +1,65 @@
+  token: string;
+  userId: string;
+  name: string;
+}
```
You have this Session interface in more files - you can move them to shared file, export it `export interface Session` and import the interface in these files.

postulkam commented on src/components/bank-account/delete-ba-button.tsx:
```diff
@@ -0,0 +1,65 @@
+  async function handleDelete() {
+    if (!confirm("Are you sure you want to delete this bank account?")) return;
+
+    setIsLoading(true);
```
Just out of interest - I reacently read about "Suspense" in React, it is interesting, look at it. :)
https://react.dev/reference/react/Suspense

"isLoading" here is ok, do not change it. :)

postulkam commented on src/components/bank-account/delete-ba-button.tsx:
```diff
@@ -0,0 +1,65 @@
+    try {
+      const response = await deleteBankAccountAction(bankAccountId, session);
+
+      if ("data" in response && response.data) {
```
Do you need it like this or can you use "response?.data" only?

postulkam commented on src/domain/bankAccount-domain/ba-actions.ts:
```diff
@@ -0,0 +1,58 @@
+    if ("data" in deletedAccount && deletedAccount.data) {
+      return successResponse("Bank account deleted successfully", deletedAccount.data);
+    } else {
+      return deletedAccount;
```
This is weird - why the else which returns only deletedAccount and no success/error response?

postulkam reviewed (commented):
This is really impresive work! I am amazed! 
Few suggestions or questions added. :)

![](https://media0.giphy.com/media/v1.Y2lkPTc5MGI3NjExd3pxb3VtbnVxbzNvaHplOXpwMnE2MjlkeTY4NWw0ZzY5ZzhtcHhvbCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/1ytPILhsu3A29ZoEl8/giphy.gif)

silviczka commented:
thank you both for insightful CR, I appreciate it! 🫶🏻😊 I' ll try to fix it all tomorrow or during weekend

silviczka committed:
Merge branch 'develop' of https://github.com/czechibank/czechibank into CZBANK-54
pull from develop

silviczka committed:
fix: backend fixes, API routes/docs, rename & soft delete

silviczka committed:
Merge remote-tracking branch 'origin/develop' into CZBANK-54
ok

silviczka commented:
Hello, 
new commit with many changes, 
BIG THANKS  to @postulkam for mentoring session it helped a lot! ❤️

Hopefully, I covered all you suggested, here short overview:
-removed duplicity
-fixed logic of connection to BE
-updated BA prisma model with isActive field for soft delete
-updated API routes and API docs
-added rename feature 
-added ba-helper fns
-added component FX (shimmer)

I'll be glad for code reviews!😊

postulkam commented on scripts/seed.ts:
```diff
@@ -205,10 +208,18 @@ function generateTransactionData(regularUsers: any[], totalTransactions: number)
-      const receiverAccount = receiver.bankAccounts[0];
-      const senderAccount = sender.bankAccounts[0];
-
+      //@Silviczka TODO clean
```
Forgoten TODO ;]

silviczka replied:
done :) thanks again for mentoring session @postulkam 🫶🏻 it helped a lot! :)

postulkam commented on src/app/api/v1/bank-account/[id]/route.ts:
```diff
@@ -76,21 +91,82 @@ import { ApiError, handleErrors } from "../../routes";
  *           application/json:
  *             schema:
  *               $ref: '#/components/schemas/Error'
+ *       401:
```
Do we get 200 code with delete here?

silviczka replied:
sorry, here I'm not sure I understand, but maybe I have fixed it in process of other chores, can you pls check if it is ok, and if not, what do I need to focus on, pls?

postulkam replied:
Yes, you can add it to another PR. 

The comment was about the '200' code in swagger documentation - there are 400, 401, etc., but no 200 code commented.

postulkam commented on src/app/api/v1/bank-account/[id]/route.ts:
```diff
@@ -146,11 +224,18 @@ export async function DELETE(request: Request, context: { params: Promise<{ id:
+    if (result.data.isActive !== false) {
+      return Response.json(
+        { success: false, message: "Bank account deletion failed", error: { code: "INTERNAL_ERROR" } },
+        { status: 500 },
```
I think, 409 will be better - the request is valid, but there is "some problem"  - https://www.w3schools.com/tags/ref_httpmessages.asp

silviczka replied:
done :)

postulkam commented on src/app/api/v1/bank-account/route.ts:
```diff
@@ -105,12 +118,34 @@ export async function GET(request: Request) {
+        errorResponse("Invalid data from bank account service", ApiErrorCode.INTERNAL_ERROR, [
+          { code: ApiErrorCode.INTERNAL_ERROR, message: "Missing or invalid data property" },
+        ]),
+        { status: 500 },
```
The same as before - look at the codes,  there will be better one for this situation maybe.

silviczka replied:
done :)

postulkam commented on src/components/bank-account/create-ba-dialog.tsx:
```diff
@@ -0,0 +1,144 @@
+      });
+
+      //simulate loading delay
+      // await new Promise((res) => setTimeout(res, 1000));
```
Forgotten comment. :)

silviczka replied:
done :)

postulkam commented on src/components/bank-account/rename-ba-dialog.tsx:
```diff
@@ -0,0 +1,118 @@
+}
+
+const renameBankAccountSchema = z.object({
+  name: z.string().min(1, "Account name cannot be empty"),
```
Can we share this schema? It is in multiple components.

silviczka replied:
done :)

silviczka replied:
hopefully I've done it correctly :D

postulkam replied:
Yes, creating the schema and reusing it was exactly what i meant. :)

postulkam commented on src/components/bank-account/delete-ba-button.tsx:
```diff
@@ -0,0 +1,83 @@
+  }
+
+  return (
+    <Suspense
```
You can remove the Suspense here, the comment was only for introducing interest topic. Suspense will not help or work in this setup.

silviczka replied:
done :)

postulkam commented on src/domain/bankAccount-domain/ba-service.ts:
```diff
@@ -1,18 +1,54 @@
+//   let nextNumber = 1;
+//   while (used.has(nextNumber)) nextNumber++;
+//   return `${baseName}(${String(nextNumber).padStart(2, "0")})`;
+// }
```
Forgotten comment ;)

silviczka replied:
done :)

postulkam reviewed (changes requested):
Just few notes, really easy to solve and then we can merge it! :) 

<img src="https://media0.giphy.com/media/v1.Y2lkPWJkM2VhNTdlMWxibTg3anpjM3RvNXB2ZXJubXZwbW0xajZrNmx0bXN1ZzZmOHlnZyZlcD12MV9naWZzX3NlYXJjaCZjdD1n/U2O50cAkpmTjG/giphy.gif"/>

silviczka committed:
chore: integration of api-error-status-map and NON_ZERO_BALANCE error

silviczka committed:
chore: show only active BAs in transfer, add 2 ApiErrorCodes, update docs/statuses

postulkam reviewed (approved):
<img src="https://media4.giphy.com/media/v1.Y2lkPWJkM2VhNTdldzR0b2NpcThzeWNldmo4MHQzMmVheHNta2F3OHFneDA3eWU1NjNxcyZlcD12MV9naWZzX3NlYXJjaCZjdD1n/fnK0jeA8vIh2QLq3IZ/giphy.gif"/>

Perfect! You can merge it. :)
